File: examples/swift/ExerciseDB.swift
// ExerciseDB API client — Swift 5.7+ (Foundation only, async/await).
//
// On Linux, add: import FoundationNetworking

import Foundation

private let baseURL = "https://api.harshitbishnoi.dev"
private let pageSize = 100

/// Errors are RFC 9457 problem+json. `code` is stable; `detail` is prose.
struct ApiError: Error, Decodable {
    let status: Int
    let code: String
    let detail: String
}

struct Exercise: Decodable {
    let id: String
    let slug: String
    let name: String
    var status: String
}

struct Tombstone: Decodable {
    let exerciseId: String
    let changeType: String
}

struct Pagination: Decodable {
    let nextCursor: String?
    let hasMore: Bool
}

struct SyncPage: Decodable {
    let exercises: [Exercise]
    let tombstones: [Tombstone]
    /// Null only when the catalog has never recorded a change.
    let latestChangeAt: String?
}

private struct Envelope<T: Decodable>: Decodable {
    let data: T
    let pagination: Pagination?
}

struct ExerciseDBClient {
    let apiKey: String
    private let session = URLSession.shared
    private let decoder = JSONDecoder()

    private func get<T: Decodable>(_ path: String) async throws -> Envelope<T> {
        var request = URLRequest(url: URL(string: baseURL + path)!)
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")

        let (data, response) = try await session.data(for: request)
        let statusCode = (response as? HTTPURLResponse)?.statusCode ?? 0

        guard (200..<300).contains(statusCode) else {
            throw try decoder.decode(ApiError.self, from: data)
        }

        return try decoder.decode(Envelope<T>.self, from: data)
    }

    func listExercises(limit: Int = 5) async throws -> [Exercise] {
        let envelope: Envelope<[Exercise]> = try await get("/exercises?limit=\(limit)")
        return envelope.data
    }

    /// Walks every page of a sync and applies it to `store`.
    ///
    /// `limit` bounds change events, not exercises: an exercise created and then
    /// updated inside the window is one record from two events, and deleted
    /// records are dropped entirely. So follow the cursor, never `count`.
    func sync(into store: LocalStore) async throws {
        let since = store.watermark
        var cursor: String?
        var watermark = since
        var pages: [SyncPage] = []

        repeat {
            var query = "limit=\(pageSize)"
            if let since { query += "&updated_since=\(since)" }
            if let cursor { query += "&cursor=\(cursor)" }

            let envelope: Envelope<SyncPage> = try await get("/sync/exercises?\(query)")
            pages.append(envelope.data)

            // Identical on every page of one sync: the server reads it before
            // the first page, so a record written mid-sync arrives on the next
            // run instead of being skipped.
            watermark = envelope.data.latestChangeAt ?? watermark
            cursor = envelope.pagination?.nextCursor
        } while cursor != nil

        // Records and watermark must commit together. A watermark that lands
        // without its records means the next sync starts after data you never
        // wrote.
        store.transaction {
            for page in pages {
                for exercise in page.exercises {
                    store.upsert(exercise)
                }

                for tombstone in page.tombstones {
                    if tombstone.changeType == "deleted" {
                        store.remove(tombstone.exerciseId)
                    } else {
                        store.markDeprecated(tombstone.exerciseId)
                    }
                }
            }

            store.watermark = watermark
        }
    }
}

/// Stand-in for your real database.
final class LocalStore {
    private(set) var exercises: [String: Exercise] = [:]
    var watermark: String?

    func upsert(_ exercise: Exercise) { exercises[exercise.id] = exercise }
    func remove(_ id: String) { exercises.removeValue(forKey: id) }

    func markDeprecated(_ id: String) {
        exercises[id]?.status = "deprecated"
    }

    func transaction(_ work: () -> Void) { work() }
}

// Usage:
//
//   let client = ExerciseDBClient(apiKey: ProcessInfo.processInfo
//       .environment["EXERCISEDB_API_KEY"] ?? "")
//
//   do {
//       for exercise in try await client.listExercises(limit: 3) {
//           print("\(exercise.slug) — \(exercise.name)")
//       }
//       try await client.sync(into: LocalStore())
//   } catch let error as ApiError where error.code == "RATE_LIMIT_EXCEEDED" {
//       print("Daily quota exhausted. Retry after midnight UTC.")
//   }

File: examples/kotlin/ExerciseDb.kt
// ExerciseDB API client — Kotlin (JVM 11+), java.net.http + kotlinx.serialization.
//
// build.gradle.kts:
//   plugins { kotlin("plugin.serialization") version "2.0.0" }
//   dependencies {
//     implementation("org.jetbrains.kotlinx:kotlinx-serialization-json:1.7.3")
//   }

package dev.exercisedb.example

import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

private const val BASE_URL = "http://localhost:3000"
private const val PAGE_SIZE = 100

/** Errors are RFC 9457 problem+json. `code` is stable; `detail` is prose. */
@Serializable
data class Problem(val status: Int, val code: String, val detail: String)

class ApiException(val problem: Problem) : RuntimeException(problem.detail)

@Serializable
data class Exercise(
    val id: String,
    val slug: String,
    val name: String,
    var status: String
)

@Serializable
data class Tombstone(val exerciseId: String, val changeType: String)

@Serializable
data class Pagination(val nextCursor: String? = null, val hasMore: Boolean)

@Serializable
data class SyncPage(
    val exercises: List<Exercise>,
    val tombstones: List<Tombstone>,
    /** Null only when the catalog has never recorded a change. */
    val latestChangeAt: String? = null
)

@Serializable
data class Envelope<T>(val data: T, val pagination: Pagination? = null)

class ExerciseDbClient(private val apiKey: String) {
    private val http: HttpClient = HttpClient.newHttpClient()
    private val json = Json { ignoreUnknownKeys = true }

    private fun fetch(path: String): String {
        val request = HttpRequest.newBuilder()
            .uri(URI.create("$BASE_URL$path"))
            .header("x-api-key", apiKey)
            .GET()
            .build()

        val response = http.send(request, HttpResponse.BodyHandlers.ofString())

        if (response.statusCode() >= 400) {
            throw ApiException(json.decodeFromString<Problem>(response.body()))
        }

        return response.body()
    }

    fun listExercises(limit: Int = 5): List<Exercise> =
        json.decodeFromString<Envelope<List<Exercise>>>(
            fetch("/exercises?limit=$limit")
        ).data

    /**
     * Walks every page of a sync and applies it to [store].
     *
     * `limit` bounds change events, not exercises: an exercise created and then
     * updated inside the window is one record from two events, and deleted
     * records are dropped entirely. So follow the cursor, never `size`.
     */
    fun sync(store: LocalStore) {
        val since = store.watermark
        var cursor: String? = null
        var watermark = since
        val pages = mutableListOf<SyncPage>()

        do {
            val query = buildString {
                append("limit=$PAGE_SIZE")
                since?.let { append("&updated_since=$it") }
                cursor?.let { append("&cursor=$it") }
            }

            val page = json.decodeFromString<Envelope<SyncPage>>(
                fetch("/sync/exercises?$query")
            )
            pages.add(page.data)

            // Identical on every page of one sync: the server reads it before
            // the first page, so a record written mid-sync arrives on the next
            // run instead of being skipped.
            watermark = page.data.latestChangeAt ?: watermark
            cursor = page.pagination?.nextCursor
        } while (cursor != null)

        // Records and watermark must commit together. A watermark that lands
        // without its records means the next sync starts after data you never
        // wrote.
        store.transaction {
            for (page in pages) {
                page.exercises.forEach(store::upsert)

                for (tombstone in page.tombstones) {
                    if (tombstone.changeType == "deleted") {
                        store.remove(tombstone.exerciseId)
                    } else {
                        store.markDeprecated(tombstone.exerciseId)
                    }
                }
            }

            store.watermark = watermark
        }
    }
}

/** Stand-in for your real database. */
class LocalStore {
    val exercises = mutableMapOf<String, Exercise>()
    var watermark: String? = null

    fun upsert(exercise: Exercise) {
        exercises[exercise.id] = exercise
    }

    fun remove(id: String) {
        exercises.remove(id)
    }

    fun markDeprecated(id: String) {
        exercises[id]?.status = "deprecated"
    }

    fun transaction(work: () -> Unit) = work()
}

fun main() {
    val apiKey = System.getenv("EXERCISEDB_API_KEY")

    if (apiKey.isNullOrEmpty()) {
        System.err.println("Set EXERCISEDB_API_KEY")
        return
    }

    val client = ExerciseDbClient(apiKey)

    try {
        client.listExercises(limit = 3).forEach { println("${it.slug} — ${it.name}") }
    } catch (error: ApiException) {
        if (error.problem.code == "RATE_LIMIT_EXCEEDED") {
            System.err.println("Daily quota exhausted. Retry after midnight UTC.")
            return
        }
        throw error
    }

    client.sync(LocalStore())
}
